alf/layers.py

```diff
@@ -733,6 +733,159 @@ def bias(self):
         return self._bias
 
 
+@alf.configurable
+class CompositionalFC(nn.Module):
+    """Compositional FC layer."""
+
+    def __init__(self,
+                 input_size,
+                 output_size,
+                 n,
+                 activation=identity,
+                 use_bias=True,
+                 use_bn=False,
+                 use_ln=False,
+                 kernel_initializer=None,
+                 kernel_init_gain=1.0,
+                 bias_init_value=0.0):
+        """
+        It maintains a set of ``n`` FC parameters for learning. During forward
+        computation, it composes the set of parameters using weighted average
+        with the compositional weight provided as input and then performs the
+        FC computation, which is equivalent to combine the pre-activation output
+        from each of the ``n`` FC layers using the compositional weight, and
+        then apply normalization and activation.
+
+        Args:
+            input_size (int): input size
+            output_size (int): output size
+            n (int): the size of the paramster set
+            activation (torch.nn.functional):
+            use_bn (bool): whether use Batch Normalization.
+            use_ln (bool): whether use layer normalization
+            use_bias (bool): whether use bias
+            kernel_initializer (Callable): initializer for the FC layer kernel.
+                If none is provided a ``variance_scaling_initializer`` with gain
+                as ``kernel_init_gain`` will be used.
+            kernel_init_gain (float): a scaling factor (gain) applied to
+                the std of kernel init distribution. It will be ignored if
+                ``kernel_initializer`` is not None.
+            bias_init_value (float): a constant
+        """
+        super().__init__()
+        self._activation = activation
+        self._weight = nn.Parameter(torch.Tensor(n, output_size, input_size))
+        if use_bias:
+            self._bias = nn.Parameter(torch.Tensor(n, output_size))
+        else:
+            self._bias = None
+
+        self._kernel_initializer = kernel_initializer
+        self._kernel_init_gain = kernel_init_gain
+        self._bias_init_value = bias_init_value
+        self._use_bias = use_bias
+        self._use_bn = use_bn
+        self._use_ln = use_ln
+        self._n = n
+
+        if use_bn:
+            self._bn = nn.BatchNorm1d(output_size)
+        else:
+            self._bn = None
+        if use_ln:
+            self._ln = nn.LayerNorm(output_size)
+        else:
+            self._ln = None
+        self.reset_parameters()
+
+    def forward(self, inputs, comp_weight=None):
```

Comment:
It might be worth to it easier to composite multiple CompositeFC using nn.Sequential. For example, FCBatchEnsemble takes a tuple as input, and also re-output emsemble_id so it can be used by the next layer in a Sequential

Reply:
Good point. Changed to support better chaining and added test.